x/mongo/driver/topology/server.go

```diff
@@ -116,12 +116,15 @@ type Server struct {
 	// globalCtx should be created in NewServer and cancelled in Disconnect to signal that the server is shutting down.
 	// heartbeatCtx should be used for individual heartbeats and should be a child of globalCtx so that it will be
 	// cancelled automatically during shutdown.
-	heartbeatLock      sync.Mutex
-	conn               *connection
-	globalCtx          context.Context
-	globalCtxCancel    context.CancelFunc
-	heartbeatCtx       context.Context
-	heartbeatCtxCancel context.CancelFunc
+	heartbeatLock   sync.Mutex
+	conn            *connection
```

Comment:
The use of `heartbeatLock` to guard `conn` is very difficult to reason about and seems inconsistent. I realize that many of the uses of `heartbeatLock` were not added in this PR, but the updates here highlight the existing problems.

Is there a way to change the heartbeat logic to guarantee that `conn` is only accessed by a single goroutine? The only concurrent access seems to come from `cancelCheck`. Can we use `checkCanceled` as an indicator that we need to cancel the check and create a new `conn` instead of accessing `conn` directly? If not, is there a way to simplify the access pattern so it's easier to understand?